Update CHANGELOG.md, update release scripts

alamb opened the pull request:
# Which issue does this PR close?

Re https://github.com/apache/arrow-datafusion/issues/1587

You can see a rendered version here: https://github.com/alamb/arrow-datafusion/blob/alamb/changelog_for_7.0.0/datafusion/CHANGELOG.md


 # Rationale for this change
Trying to tell the 🌎  🌍  🌏   about DataFusion

# What changes are included in this PR?
It was created using

```shell
$ ./dev/release/update_change_log-datafusion.sh
```

🤯  what a lot of stuff happened!!!

Since this file is automatically generated: to make changes, please edit the ticket subjects / labels directly (or tag me `@alamb` if you don't have the permissions to do so).

I am hoping to create a release candidate over the weekend (as I will be largely offline starting next Thursday Feb 17 for a week or so) so I can do the release next week before I head out.

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

alamb commented on dev/release/update_change_log.sh:
```diff
@@ -50,14 +50,14 @@ OUTPUT_PATH="${PROJECT}/CHANGELOG.md"
 
 # reset content in changelog
-git co "${SINCE_TAG}" "${OUTPUT_PATH}"
+git checkout "${SINCE_TAG}" "${OUTPUT_PATH}"
```
my mac didn't like `git co` 😢

houqp replied:
haha, sorry that's an alias i made up in my gitconfig, muscle memory...

alamb replied:
No problem. I tip my hat to you for these scripts in general (automating the RAT in particular is genius, I normally just fix it up manually in arrow-rs)

alamb commented on dev/release/update_change_log.sh:
```diff
@@ -50,14 +50,14 @@ OUTPUT_PATH="${PROJECT}/CHANGELOG.md"
+git checkout "${SINCE_TAG}" "${OUTPUT_PATH}"
 # remove license header so github-changelog-generator has a clean base to append
-sed -i '1,18d' "${OUTPUT_PATH}"
+sed -i.bak '1,18d' "${OUTPUT_PATH}"
```
Likewise, apparently `sed` from mac is slightly different than linux 

https://stackoverflow.com/questions/5694228/sed-in-place-flag-that-works-both-on-mac-bsd-and-linux

alamb commented on dev/release/update_change_log.sh:
```diff
@@ -50,14 +50,14 @@ OUTPUT_PATH="${PROJECT}/CHANGELOG.md"
     -e CHANGELOG_GITHUB_TOKEN=$CHANGELOG_GITHUB_TOKEN \
     -v "$(pwd)":/usr/local/src/your-app \
-    githubchangeloggenerator/github-changelog-generator:1.16.2 \
+    githubchangeloggenerator/github-changelog-generator \
```
use latest

houqp replied:
might want to pin to the latest version number instead to avoid breaking changes in the future.

alamb replied:
🤔  https://hub.docker.com/r/githubchangeloggenerator/github-changelog-generator/tags

Seems like `latest` has no other tag (as in `latest` is newer than `1.16.2` but there are no other numbered versions newer than `1.16.2`) 😞 

![Screen Shot 2022-02-11 at 6 00 42 AM](https://user-images.githubusercontent.com/490673/153580452-54ccec0a-47bb-48b2-b5e9-ec7967d58920.png)

xudong963 commented on datafusion/CHANGELOG.md:
```diff
@@ -19,6 +19,293 @@
+- Simplify creating new `ListingTable`  [\#1705](https://github.com/apache/arrow-datafusion/issues/1705)
+- Implement TableProvider for DataFrameImpl to allow registration of logical plans [\#1698](https://github.com/apache/arrow-datafusion/issues/1698)
+- Public Expr simplification API [\#1694](https://github.com/apache/arrow-datafusion/issues/1694)
+- Query Optimizer: Add OUTER --\> INNER join conversion [\#1670](https://github.com/apache/arrow-datafusion/issues/1670)
```
In fact, this is not implemented. BTW, I am worried that there are some errors similar to this one, we just closed repeated issues, but features are not implemented.

xudong963 replied:
I did not explore this script, which can generate changelog based on PRs already merged?

xudong963 commented on datafusion/CHANGELOG.md:
```diff
@@ -19,6 +19,293 @@
+-  Support DataType::Decimal\(15, 2\) in TPC-H benchmark [\#174](https://github.com/apache/arrow-datafusion/issues/174)
+-   Make `MemoryStream` public [\#150](https://github.com/apache/arrow-datafusion/issues/150)
+-   Add support for Parquet schema merging  [\#132](https://github.com/apache/arrow-datafusion/issues/132)
+-   Add SQL support for IN expression [\#118](https://github.com/apache/arrow-datafusion/issues/118)
```
ditto

xudong963 reviewed (commented):
Thanks @alamb . I gave some suggestions.

houqp reviewed (commented):
automation fix looks good to me :+1: looks like we still have a bit of work left to update PRs with proper tags :D

matthewmturner commented:
I'm interested in learning a bit more about the lifecycle of issues / tags / releases.  

@houqp @alamb could you provide a little more info? i didnt see anything mentioned in the developers guide.

once i understand better maybe i could help to get things in order for the next release.

alamb commented:
> I'm interested in learning a bit more about the lifecycle of issues / tags / releases.
> once i understand better maybe i could help to get things in order for the next release.

🎉  that would be wonderful

> @houqp @alamb could you provide a little more info? i didnt see anything mentioned in the developers guide.

The basic release instructions https://github.com/apache/arrow-datafusion/blob/master/dev/release/README.md#update-changelogmd

Talk about 
```shell
# create the changelog
CHANGELOG_GITHUB_TOKEN=<TOKEN> ./dev/release/update_change_log-all.sh
# review change log / edit issues and labels if needed, rerun until you are happy with the result
git commit -a -m 'Create changelog for release'
```

But the "until you are happy with the result" leaves a lot to the imagination 😆 

Basically what I did was run that script (it updates the `CHANGELOG.md` file locally) and then looked at the output and tried to make something that looked coherent.

Examples of things that I did:
1. Found issues that did not have `datafusion` but were related to datafusion and put them in (otherwise they don't show up in the release notes)
2. Changed  titles of PRs / issues so they were more specified (e.g. from `Fixed bug in select` to `Fixed bug when there are order by number`)
3. Applied a liberal dose of judgement to what issues / PRs should have `enhancement` `bug` or `api-change` on them. 

The most questionable thing related to labels were:
1. API changes get listed under the "Beaking Changes" labels, and some tickets had new apis, etc that I didn't feel were breaking changes or if there were several PRs that together made a "single" breaking change from the user's point of view (e.g. breaking LogicalPlan into enums)
2. Changes with multiple labels got put under a single heading so I removed some labels that were accurate but were obscuring what I felt was the "most important" part. For example, a ticket with an `enhancement` and a `documentation` label ended up under the Documentation heading, even when it also had code changes. I removed the `documentation` ticket for that one

It would be great to have some more help here. Some thoughts:
1. We could probably automate adding `datafusion` labels for issues that are closed via PR that also has the `datafusion` label (and the `datafusion` label is applied automatically based on the path that was fixed)

Something else I haven't even been looking into is releasing the python bindings and releasing ballista lol -- so any help you want to lend in there would be awesome
